src/index-file/index-file.ts

```diff
@@ -9,41 +9,75 @@ import {
   readIndexMeta,
 } from "./meta";
 import { FieldType } from "../db/database";
+import { parseMultipartBody } from "../data-file";
 
 export class IndexFile {
   static async forUrl<T = any>(url: string) {
-    return await IndexFile.forResolver<T>(
-      async ({ start, end, expectedLength }) => {
-        const response = await fetch(url, {
-          headers: { Range: `bytes=${start}-${end}` },
+    return await IndexFile.forResolver<T>(async (ranges) => {
+      const rangesHeader = ranges
+        .map(({ start, end }) => `${start}-${end}`)
+        .join(",");
+
+      const response = await fetch(url, {
+        headers: {
+          Range: `bytes=${rangesHeader}`,
+          Accept: "multipart/bytesranges",
+        },
+      });
+
+      if (response.status !== 206) {
+        throw new Error(`Expected 206 Partial Content, got ${response.status}`);
+      }
+
+      const contentType = response.headers.get("Content-Type");
+      if (!contentType) {
+        throw new Error("Missing Content-Type in response");
+      }
+      if (contentType.includes("multipart/byteranges")) {
+        const boundary = contentType.split("boundary=")[1];
+        const abuf = await response.arrayBuffer();
+        const text = new TextDecoder().decode(abuf);
+
+        const chunks = parseMultipartBody(text, boundary);
+
+        // the last element is null since the final boundary marker is followed by another delim.
+        if (chunks[chunks.length - 1].body === undefined) {
+          chunks.pop();
+        }
+        console.log(chunks);
+        return chunks.map((c) => {
+          const enc = new TextEncoder();
+          const data = enc.encode(c.body).buffer;
+
+          const totalLengthStr = c.headers["content-range"]?.split("/")[1];
+          const totalLength = totalLengthStr ? parseInt(totalLengthStr, 10) : 0;
+
+          return { data, totalLength };
         });
+      } else if (response.headers.has("Content-Range")) {
+        const abuf = await response.arrayBuffer();
         const totalLength = Number(
-          response.headers.get("Content-Range")!.split("/")[1]
+          response.headers.get("Content-Range")!.split("/")[1],
         );
-        if (expectedLength && totalLength !== expectedLength) {
-          throw new LengthIntegrityError();
-        }
-        return {
-          data: await response.arrayBuffer(),
-          totalLength: totalLength,
-        };
+        return [
+          {
+            data: abuf,
+            totalLength: totalLength,
+          },
+        ];
```

Comment:
Can this be put into a shared function with `DataFile`? Maybe create a `range-requests.ts` and export the function to parse the response.